B2C Implementation Guide (#7310)
auth0/docs: articles/architecture-scenarios/implementation/b2c/architecture.md

::: warning
Tenant names cannot be changed or reused once deleted. So, make sure you're happy with your name(s) before you create your Auth0 tenants.
:::

Determining the level of isolation you require when it comes to your user domains is an important step, and together with your branding requirements will subsequently help you determine the number of Auth0 tenants that will be required in your production environment. As we recommend you create a full suite of SDLC supporting tenants for every Auth0 tenant you will run in a production environment the number of Auth0 tenants you will need to manage can quickly grow. Therefore you should consider carefully before creating multiple Auth0 tenants for production, and should consult our guidance on [Branding](/architecture-scenarios/implementation/b2c/branding) before making your final decision. 

## Custom domains

> When you setup your Auth0 tenant, the URL for accessing that tenant will be `https://${account.tenant}.auth0.com`. A custom domain, also known as a [vanity URL](/custom-domains), for your Auth0 tenant is important for more than just branding. It also provide you with security benefits too:

Reviewer: ```suggestion
When you setup your Auth0 tenant, the URL for accessing that tenant will be `https://${account.tenant}.auth0.com`. Providing a custom domain, also known as a [vanity URL](/custom-domains), for your Auth0 tenant is an important factor for supporting your Branding requirements, but more importantly will also provide you with security benefits too:
```

* Some browsers will, by default, make it difficult to communicate in an iFrame if you don't have a shared domain. (Not the case however if you have multiple domain SSO.)

> * It is harder to phish your domain if you have a vanity URL. The phisher must also create a vanity URL to mimic your. For example, with a custom domain, you can use your own certificate and get an "Extended Validation Certificate" to make phishing even harder.

Reviewer: ```suggestion
* It's harder to phish your domain if you have a vanity URL. The phisher must also create a vanity URL to mimic your. For example, with a custom domain, you can use your own certificate to get an "Extended Validation", making phishing even harder.
```

* You want your users to be trained to look for suspicious URLs when entering their passwords. **This training is extremely important.**

::: note
You are allowed only one custom domain per Auth0 Tenant. This is because a tenant in Auth0 is intended to represent a “domain” of users. If you need more than one vanity URL, then you likely have more than one domain of users and should be using multiple tenants.
:::

> Create your custom domain in all environments early on to ensure that you are testing consistently between environments. Your custom domain name should also give the user confidence that this is the appropriate place to enter their credentials.

Reviewer: ```suggestion
Your custom domain name should also give the user confidence that this is the appropriate place to enter their credentials, and we recommend that you create your custom domain in all environments early on to ensure that you are testing consistently between environments.
```

## SDLC support

Every company has some form of software development life cycle. Throughout the development process, we need to be able to test the authorization service integration as well as the applications themselves. It is therefore important to structure Auth0 tenants to support this life cycle, and there tends to be consistency around the best practices associated with tenant layout for supporting the SDLC:

| Environment | Sample Tenant Name | Description |
| - | - | - |
| Development | **company-dev** | A shared environment where most of your development work occurs |
| QA/Testing | **company-qa** or **company-uat** | An environment for formal testing of the changes you've made |
| Production | **company-prod** | The production tenant |

You may also want to create one or more sandboxes (e.g., **company-sandbox1**, **company-sandbox2**) so that you can test changes without compromising your development environment. This might be where you test deployment scripts and the like.

::: warning
Though Auth0 allows you to create as many free tenants as you'd like, you can only have **three** tenants with paid features enabled.
:::

## Tenant association

> To ensure that your [tenants are all associated with your Auth0 contractual agreement](/dev-lifecycle/child-tenants) and have the same features, ensure you have all of your tenants associated with your company. If you have individual developers that want to create their own sandboxes for testing, make sure they get associated with your account so they have the same permissions. To do this you should contact your Auth0 representative or the Auth0 Support Center at ${env.DOMAIN_URL_SUPPORT}.

Reviewer: ```suggestion
To ensure that your [tenants are all associated with your Auth0 contractual agreement](/dev-lifecycle/child-tenants) and have the same features, ensure you have all of your tenants are associated with your company account. If you have individual developers that want to create their own sandboxes for testing, make sure they get associated with your account so they have the same permissions too. To do this you should contact your Auth0 representative or the Auth0 Support Center at ${env.DOMAIN_URL_SUPPORT}.
```
